src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs

```diff
@@ -0,0 +1,79 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.CommandLine;
+using System.CommandLine.Builder;
+using System.CommandLine.Parsing;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.Interactive.Parsing
+{
+    public class DirectiveParseResult
+    {
+        private ParseResult _commandLineParseResult;
+        public ParseResult ParseResult => _commandLineParseResult;
+        public DirectiveParser Parser { get; }
+        public DirectiveParseResult(ParseResult commandLineParseResult, DirectiveParser parser)
+            => (Parser, _commandLineParseResult) = (parser, commandLineParseResult);
+
+        public IReadOnlyCollection<ParseError> Errors { get; init; }
+        public IReadOnlyList<string> UnmatchedTokens { get; init; }
+        public IReadOnlyList<string> UnparsedTokens => _commandLineParseResult.UnparsedTokens;
+        public IReadOnlyList<Token> Tokens => _commandLineParseResult.Tokens;
+        public CommandResult CommandResult => _commandLineParseResult.CommandResult;
+
+        public IEnumerable<string> GetSuggestions(int requestPosition) => _commandLineParseResult.GetSuggestions(requestPosition);
+        public async Task<int> InvokeAsync() => await _commandLineParseResult.InvokeAsync();
+        public T ValueForArgument<T>(string name) => _commandLineParseResult.ValueForArgument<T>(name);
+    }
+
+    public class DirectiveParser
+    {
+        private readonly Parser _commandLineParser;
+        public DirectiveParser(RootCommand rootCommand)
+        {
+            var commandLineBuilder =
+                    new CommandLineBuilder(rootCommand)
+                        .ParseResponseFileAs(ResponseFileHandling.Disabled)
+                        .UseTypoCorrections()
+                        .UseHelpBuilder(bc => new DirectiveHelpBuilder(rootCommand.Name))
+                        .UseHelp()
+                        .UseMiddleware(
+                            context =>
+                            {
+                                context.BindingContext
+                                       .AddService(
+                                           typeof(KernelInvocationContext),
+                                           _ => KernelInvocationContext.Current);
+                            });
+            commandLineBuilder.EnableDirectives = false;
+            _commandLineParser = commandLineBuilder.Build();
+        }
+
+        public DirectiveParseResult Parse(string input)
+        {
+            var parseResult = _commandLineParser.Parse(input);
+            IReadOnlyList<string> unmatchedTokens;
+            IReadOnlyCollection<ParseError> errors;
+            if (parseResult.UnmatchedTokens.Count == parseResult.Errors.Count && parseResult.UnmatchedTokens.Count > 0 && parseResult.UnmatchedTokens[0].StartsWith("//"))
```

Comment:
@jonsequitur is it a bad solution, or *very* bad? By this trick when I get as many errors as unmatched tokens, all those errors look like `Unrecognized token bla-bla`, so we can check if the first token starts with `//`.

Reply:
Interesting solution. It's not too, too bad, given the alternative would probably be to introduce a whole other parser.

This could be tightened up further by making sure that the unmatched tokens all come after the matched tokens. It's possible, I think, for there to be unmatched tokens in between matched tokens.

Reply:
Aha, I see. Alright, then I'll tighten it up and add tests.